source/host_resolver.c

```diff
@@ -732,52 +732,79 @@ static int default_resolve_host(
-            AWS_LOGF_TRACE(
-                AWS_LS_IO_DNS,
-                "id=%p: vending address %s for host %s to caller",
-                (void *)resolver,
-                aaaa_record->address->bytes,
-                host_entry->host_name->bytes);
+            struct aws_host_address aaaa_record_cpy;
+            if (!aws_host_address_copy(aaaa_record, &aaaa_record_cpy)) {
+                aws_array_list_push_back(&callback_address_list, &aaaa_record_cpy);
+                AWS_LOGF_TRACE(
+                    AWS_LS_IO_DNS,
+                    "id=%p: vending address %s for host %s to caller",
+                    (void *)resolver,
+                    aaaa_record->address->bytes,
+                    host_entry->host_name->bytes);
+            }
         }
         if (a_record) {
-            aws_array_list_push_back(&callback_address_list, &a_record);
-            AWS_LOGF_TRACE(
-                AWS_LS_IO_DNS,
-                "id=%p: vending address %s for host %s to caller",
-                (void *)resolver,
-                a_record->address->bytes,
-                host_entry->host_name->bytes);
+            struct aws_host_address a_record_cpy;
+            if (!aws_host_address_copy(a_record, &a_record_cpy)) {
+                aws_array_list_push_back(&callback_address_list, &a_record_cpy);
+                AWS_LOGF_TRACE(
+                    AWS_LS_IO_DNS,
+                    "id=%p: vending address %s for host %s to caller",
+                    (void *)resolver,
+                    a_record->address->bytes,
+                    host_entry->host_name->bytes);
+            }
         }
+        aws_rw_lock_wunlock(&host_entry->entry_lock);
+        aws_rw_lock_runlock(&default_host_resolver->host_lock);
 
-        res(host_entry->resolver, host_entry->host_name, AWS_OP_SUCCESS, &callback_address_list, user_data);
-    } else {
-        struct pending_callback *pending_callback =
-            aws_mem_acquire(default_host_resolver->allocator, sizeof(struct pending_callback));
-        pending_callback->user_data = user_data;
-        pending_callback->callback = res;
-        aws_linked_list_push_back(&host_entry->pending_resolution_callbacks, &pending_callback->node);
+        int error_code = AWS_OP_SUCCESS;
+        /* we don't want to do the callback WHILE we hold the lock someone may reentrantly call us. */
+        if (aws_array_list_length(&callback_address_list)) {
+            res(resolver, host_name, AWS_OP_SUCCESS, &callback_address_list, user_data);
+        } else {
+            res(resolver, host_name, aws_last_error(), NULL, user_data);
+            error_code = AWS_OP_ERR;
+        }
+
+        for (size_t i = 0; i < aws_array_list_length(&callback_address_list); ++i) {
+            struct aws_host_address *address_ptr = NULL;
+            aws_array_list_get_at_ptr(&callback_address_list, (void **)&address_ptr, i);
+            aws_host_address_clean_up(address_ptr);
```

Comment:
[long rant] I really dislike using array_list when the size is known. If we used C-arrays we'd get code that's much easier to read and write, plus we'd get type-checking by the compiler.

I was going to say how it's too late to change this because the array_list is passed to a callback that is already part of the public API. But that brings up a good point. These changes SHOULD have broken the public API. The array_list now holds pointers, it used to hold values, but the API didn't change. **And the comments on `aws_on_host_resolved_result_fn` still talk about how the array_list holds things by value**.

So we DID break the API, but a user who took this change would not be alerted of the break by the compiler, so they'd have to chase down bugs at runtime instead, because array_lists are not type-safe, in addition to being harder to use when there was no need to be dynamic anyway.

Reply:
well we'd also have to pass number of entries, but I guess that's not the end of the world. I can do that I guess.

Reply:
hmmmm, the more I think about it the more I think this isn't the case. Theres nothing guaranteeing how many records we'll return in the future. Could be all of them in the case of multi-answer. It's just an implementation detail at the moment that we know the max size in advance.

Reply:
so the underlying piece that calls getaddrinfo, returns a variable length list because we don't know how many will come back. Could be 1, could be 24

Reply:
If we ever want to change the load balancing behavior in channel bootstrap, we'd have to go back and convert back to array list anyways.

Reply:
and it was documented as by-value, but in reality it was by reference, now it's actually by-value again. this was because we had to copy to avoid the race condition.

Reply:
If you pass the array and its length to the callback, there's nothing stopping you from using an array_list in the implementation. You can pass the buffer and length to the callback.